Source/JavaScriptCore/wasm/WasmB3IRGenerator.cpp

```diff
@@ -2324,6 +2327,109 @@ auto B3IRGenerator::addI31GetU(ExpressionType ref, ExpressionType& result) -> Pa
     return { };
 }
 
+auto B3IRGenerator::addStructNew(uint32_t typeIndex, Vector<ExpressionType>& args, ExpressionType& result) -> PartialResult
+{
+    const auto type = Type { TypeKind::Ref, Nullable::No, m_info.typeSignatures[typeIndex]->index() };
+    PatchpointValue* patch = m_proc.add<PatchpointValue>(toB3Type(type), origin());
+    patch->append(instanceValue(), ValueRep::reg(GPRInfo::argumentGPR0));
+    patch->append(framePointer(), ValueRep::reg(GPRInfo::argumentGPR1));
+    patch->clobber(RegisterSet::volatileRegistersForJSCall());
+    patch->resultConstraints = Vector<B3::ValueRep, 1> { ValueRep::reg(GPRInfo::returnValueGPR) };
+    for (unsigned i = 0; i < args.size(); ++i)
+        patch->append(get(args[i]), ValueRep::stackArgument(i * sizeof(EncodedJSValue)));
+
+    PatchpointExceptionHandle handle = preparePatchpointForExceptions(m_currentBlock, patch);
+    patch->setGenerator([this, typeIndex, handle] (CCallHelpers& jit, const B3::StackmapGenerationParams& params) {
+        AllowMacroScratchRegisterUsage allowScratch(jit);
+        handle.generate(jit, params, this);
+        emitStructNewImpl(jit, typeIndex);
+    });
+    m_currentBlock->append(patch);
+    result = push(patch);
+
+    return { };
+}
+
+auto B3IRGenerator::addStructGet(ExpressionType structReference, uint32_t typeIndex, uint32_t fieldIndex, ExpressionType& result) -> PartialResult
+{
+    const auto& structType = *m_info.typeSignatures[typeIndex]->as<StructType>();
+    ASSERT(typeIndex < m_info.typeCount());
+    ASSERT(fieldIndex < structType.fieldCount());
+
+    Value* resultAsInt64 = m_currentBlock->appendNew<CCallValue>(m_proc, B3::Int64, origin(),
+        m_currentBlock->appendNew<ConstPtrValue>(m_proc, origin(), tagCFunction<OperationPtrTag>(operationWasmStructGet)),
+        get(structReference), m_currentBlock->appendNew<Const32Value>(m_proc, origin(), fieldIndex));
+
+    const auto fieldType = structType.field(fieldIndex).type;
+    switch (fieldType.kind) {
+    case TypeKind::I32:
+        result = push(m_currentBlock->appendNew<Value>(m_proc, B3::Trunc, origin(), resultAsInt64));
+        break;
+    case TypeKind::Funcref:
+    case TypeKind::Externref:
+    case TypeKind::Ref:
+    case TypeKind::RefNull:
+    case TypeKind::I64:
+        result = push(resultAsInt64);
+        break;
+    case TypeKind::F32: {
+        Value* resultAsI32 = m_currentBlock->appendNew<Value>(m_proc, B3::Trunc, origin(), resultAsInt64);
+        result = push(m_currentBlock->appendNew<Value>(m_proc, B3::BitwiseCast, origin(), resultAsI32));
+        break;
+    }
+    case TypeKind::F64:
+        result = push(m_currentBlock->appendNew<Value>(m_proc, B3::BitwiseCast, origin(), resultAsInt64));
+        break;
+    default:
+        RELEASE_ASSERT_NOT_REACHED();
+    }
+
+    return { };
+}
```

Comment:
We should inline this load. This can be a follow up, but we need a bug opened for this and a FIXME that links to it.